rclpy/src/rclpy/guard_condition.hpp

```diff
@@ -17,31 +17,73 @@
 
 #include <pybind11/pybind11.h>
 
+#include <rcl/error_handling.h>
+#include <rcl/guard_condition.h>
+
+#include <memory>
+
+#include "destroyable.hpp"
+#include "rclpy_common/exceptions.hpp"
+#include "rclpy_common/handle.h"
+#include "utils.hpp"
+
 namespace py = pybind11;
 
 namespace rclpy
 {
 /// Create a general purpose guard condition
-/**
- * A successful call will return a Capsule with the pointer of the created
- * rcl_guard_condition_t * structure
- *
- * Raises RuntimeError if initializing the guard condition fails
- *
- * \return a guard condition capsule
- */
-py::capsule
-guard_condition_create(py::capsule pycontext);
-
-/// Trigger a general purpose guard condition
-/**
- * Raises ValueError if pygc is not a guard condition capsule
- * Raises RCLError if the guard condition could not be triggered
- *
- * \param[in] pygc Capsule pointing to guard condtition
- */
-void
-guard_condition_trigger(py::capsule pygc);
+class GuardCondition : public Destroyable, public std::enable_shared_from_this<GuardCondition>
+{
+public:
+  /**
+   * Raises RuntimeError if initializing the guard condition fails
+   */
+  explicit GuardCondition(py::capsule pycontext);
+
+  /// Trigger a general purpose guard condition
+  /**
+   * Raises ValueError if pygc is not a guard condition capsule
+   * Raises RCLError if the guard condition could not be triggered
+   */
+  void
+  trigger_guard_condition();
+
+  /// Get rcl_client_t pointer
+  rcl_guard_condition_t * rcl_ptr() const
+  {
+    return rcl_guard_condition_.get();
+  }
+
+  /// Handle destructor for guard condition
+  static void
+  _rclpy_destroy_guard_condition(void * p)
+  {
+    (void)p;
+    // Empty destructor, the class should take care of the lifecycle.
```

Comment:
@ahcorde what ensures this object won't get trashed before the capsule does?

Reply:
The Python class `SignalHandlerGuardCondition(GuardCondition)` is the only one in the code making use of this reference. In the destructor I can see how the capsule is destroyed first and then the object.

https://github.com/ros2/rclpy/blob/7dff0ee8bc6f299b46a6d0c2f8adc5cd518ae95f/rclpy/rclpy/signals.py#L38-L42

or should we include some kind of logic ?

Reply:
I see. This is arguably a (temporary?) implementation detail. No need to change it.